docs/proposals/13-automated-seed-management.md

```diff
@@ -0,0 +1,369 @@
+
+**Note:** Even if the controller is called *Horizontal Pod Autoscaler*, it is capable of scaling any resource with a `scale` subresource, using any custom metric. Therefore, initially it was proposed to use this controller directly. However, a number of important drawbacks were identified with this approach, and so it is no longer proposed here.
+
+### SeedAutoscaler
+
+The SeedAutoscaler automatically scales the number of [ShootedSeeds](#shootedseeds) in a [ShootedSeedSet](#shootedseedsets) based on observed resource utilization. The resource could be any resource that is tracked via the `capacity` and `allocatable` fields in the Seed status, including in particular the number of shoots already scheduled onto the seed.
+
+#### SeedAutoscaler Resource
+
+The SeedAutoscaler is implemented as a custom resource and a new controller. The resource determines the behavior of the controller. The `SeedAutoscaler` resource has a `scaleTargetRef` that specifies the target resource to be scaled, the minimum and maximum number of replicas, as well as a list of metrics. The only supported metric type initially is `Resource` for resources that are tracked via the `capacity` and `allocatable` fields in the Seed status. The resource target can be of type `Utilization` or `AverageValue`.
+
+Example `SeedAutoscaler` resource:
+
+```yaml
+apiVersion: seedmanagement.gardener.cloud/v1alpha1
+kind: SeedAutoscaler
+metadata:
+  name: crazy-botany
+  namespace: garden
+spec:
+  scaleTargetRef:
+    apiVersion: seedmanagement.gardener.cloud/v1alpha1
+    kind: ShootedSeedSet
+    name: crazy-botany
+  minReplicas: 1
+  maxReplicas: 10
+  metrics:
+  - type: Resource # Only Resource is supported
+    resource:
+      name: shoots
+      target:
+        type: Utilization # Utilization or AverageValue
+        averageUtilization: 50
+```
+
+#### SeedAutoscaler Controller
+
+`SeedAutoscaler` resources are reconciled by a new controller, either in `gardenlet` or out-of-tree, similarly to [cluster-autoscaler](https://github.com/gardener/autoscaler). The controller periodically adjusts the number of replicas in a ShootedSeedSet to match the observed average resource utilization to the target specified by user.
+
+**Note:** The SeedAutoscaler controller is not limited to evaluating only metrics, it could also evaluate taints, label selectors, etc. This is not yet reflected in the example `SeedAutoscaler` resource below. Such details are intentionally not specified in this GEP, they should be further explored in the issues created to track the actual implementation.
```

Comment:
This is also a bit far ahead, but I am also not too sure about whether this metrics approach you proposed (which is also used in the `HorizontalPodAutoscaler`) is the right one for scaling `ShootedSeedSet`s.
Once the scheduling logic got more and more complex, evaluating some metrics might not be sufficient anymore, for example for respecting the mentioned taints/tolerations and label selectors and so on.

I would rather try to align this with the logic and mechanisms of the `cluster-autoscaler`. Though, we can't only scale up on `Pending` Shoots like CA, but rather want to scale up proactively.
This will also get very complex quite fast, but still I think, it will be better approach, as the management of the available Seeds is directly bound to the scheduling policies and mechanisms.

Reply:
I am not sure I completely get your point. I agree we should not only scale based on metrics (and this is already mentioned in the document), but aren't the metrics still one of the major factors to consider for scaling? How could one go without them?
Perhaps my problem is that I don't quite get what you mean by "the management of the available Seeds is directly bound to the scheduling policies and mechanisms.", could you please clarify how this is different from a metrics-based approach?

Reply:
In an in-person discussion, we clarified that:

* The current proposal describes scaling individual `ManagedSeedSets` proactively based on their resource utilisation metrics. The cluster autoscaler is looking at a "holistic picture" of a number of machine sets and picks one out of them to be scaled reactively, whenever a pod can't be scheduled due to insufficient resources. 
* The proposed approach is a good starting point. If at any point in the future we figure out that we need to evaluate a "holistic picture" similarly to how the cluster autoscaler does it, we could enhance or change it accordingly.